returnn/tf/layers/basic.py

```diff
@@ -913,44 +913,227 @@ def transform_config_dict(cls, d, network, get_layer):
     """
-    :param str axis:
-    :param int position:
+    :param LayerBase|int position: Layer containing the indices used to select the slices of the input from.
+      If another layer, must be of type ``int32`` or ``int64``.
+      Can also specify a constant ``int``.
+    :param str axis: The axis into which we gather the indices into
     """
     super(GatherLayer, self).__init__(**kwargs)
-    axis = self.input_data.get_axis_from_description(axis, allow_int=False)
-    assert isinstance(position, int)  # nothing else implemented yet
-    slices = [slice(None, None)] * axis + [position]  # type: typing.List[typing.Union[slice, int]]
-    self.output.placeholder = self.input_data.placeholder[slices]
+    self.position = position
+
+    input_data = self.input_data
+    if isinstance(position, int):
+      position_data = Data.from_tensor(tf.constant(position))
+    else:
+      position_data = position.output
+    old_gather_axis = input_data.get_axis_from_description(axis, allow_int=False)  # might be moved later
+
+    # determine all common axes of input_data and position_data
+    common_axes_input, common_axes_position, input_axes, position_axes = (
+      self._get_common_input_position_axes(input_data, position_data, old_gather_axis))
+    batch_dims = len(common_axes_input)
+
+    # move gather axis back if necessary s.t. common axes are always in front
+    gather_axis = max(batch_dims, old_gather_axis)
+
+    # (BatchAxes.., InputAxesBeforeGatherAxis, gather_axis, InputAxesAfterGatherAxis..)
+    params = tf.transpose(
+      input_data.placeholder,
+      common_axes_input + input_axes[:gather_axis-batch_dims] + [old_gather_axis] + input_axes[gather_axis-batch_dims:])
+    # (BatchAxes.., PositionAxes..)
+    indices = tf.transpose(position_data.placeholder, common_axes_position + position_axes)
+    # (BatchAxes.., InputAxesBeforeGatherAxis, PositionAxes.., InputAxesAfterGatherAxis..)
+    self.output.placeholder = tf.gather(params=params, indices=indices, axis=gather_axis, batch_dims=batch_dims)
+
+    # Take size_placeholder from input_data if they exist there, otherwise from position_data
+    size_placeholder = {}
+    for input_axis, size in input_data.size_placeholder.items():
+      input_axis = input_data.get_batch_axis(input_axis)
+      if input_axis == old_gather_axis:
+        continue
+      output_axis = self.output.get_batch_axis_excluding_batch(
+        self._translate_input_axis(input_axis, old_gather_axis, common_axes_input, input_axes, position_axes))
+      size_placeholder[output_axis] = size
+    for position_axis, size in position_data.size_placeholder.items():
+      position_axis = position_data.get_batch_axis(position_axis)
+      output_axis = self.output.get_batch_axis_excluding_batch(self._translate_position_axis(
+        position_axis, old_gather_axis, common_axes_position, position_axes))
+      size_placeholder.setdefault(output_axis, size)
+    self.output.size_placeholder = size_placeholder
```

Comment:
The `size_placeholder` should be set in `get_out_data_from_opts`, not here in `__init__`.
(Always for existing dyn sizes. The only exception is currently for new dyn sizes, and even that will change.)